### ISSUE TITLE
Dice loss function of Torch version

### ISSUE THREAD

**Issue by Alison-brie** [voxelmorph/voxelmorph#367]
Hi,
I have some difficulties using the Dice loss function as it seems a little vague to me what exactly is its parameters.
There are two required parameters,  y_true and y_pred.
Could you please explain the desired format of these parameters?
Like what shape should be? Like (batch_size, 1, H, M, N)?
I also don't know whether the segmentation masks( y_true and y_pred) should be converted to the one-hot format?
Could I use segmentation masks obtained from Freesurfer directly to compute the dice loss without any processing?

Thanks for your help.

```
class Dice:
    """
    N-D dice for segmentation
    """

    def loss(self, y_true, y_pred):
        ndims = len(list(y_pred.size())) - 2
        vol_axes = list(range(2, ndims + 2))
        top = 2 * (y_true * y_pred).sum(dim=vol_axes)
        bottom = torch.clamp((y_true + y_pred).sum(dim=vol_axes), min=1e-5)
        dice = torch.mean(top / bottom)
        return -dice

```

**Comment by Alison-brie** [comment]
Another question is that when computing dice loss, we should apply the SpatialTransformer of 'nearest' mode to warp the segmentation masks using flow fields, which seems to not support backpropagation (more specifically, nearest-neighbor interpolation operation). So how to handle it? Any suggestion?

Thanks for the help in advance.

**Comment by Alison-brie** [comment]
Referring to solved Issues, I guess that maybe with one-not encoding, could we use a linear interpolation when warping?

**Comment by adalca** [comment]
@Alison-brie I haven't used the pytorch code as much, but I think that indeed we are looking for the one-hot (potentially linearly warped) deformations, and the is Soft Dice being computed. So it seems like you had all the answers already :D

**Comment by Alison-brie** [comment]
Okay, get it! Thanks for the response :)
